### PR TITLE
[Reassociate] Use disjoint flag to convert Or to Add.

### PR DESCRIPTION
**Pull request by topperc**
Stacked on #72702 and #72583

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/Reassociate.cpp [hunk 1]
```diff
@@ -2252,9 +2252,10 @@ void ReassociatePass::OptimizeInst(Instruction *I) {
   // with no common bits set, convert it to X+Y.
   if (I->getOpcode() == Instruction::Or &&
       shouldConvertOrWithNoCommonBitsToAdd(I) && !isLoadCombineCandidate(I) &&
-      haveNoCommonBitsSet(I->getOperand(0), I->getOperand(1),
-                          SimplifyQuery(I->getModule()->getDataLayout(),
-                                        /*DT=*/nullptr, /*AC=*/nullptr, I))) {
+      (I->isDisjoint() ||
+       haveNoCommonBitsSet(I->getOperand(0), I->getOperand(1),
+                           SimplifyQuery(I->getModule()->getDataLayout(),
+                                         /*DT=*/nullptr, /*AC=*/nullptr, I)))) {
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/Reassociate.cpp:2258]: I think we should instead infer `disjoint` in InstCombine and then replace all existing `haveNoCommonBitsSet()` checks with only a `isDisjoint()` check, not both.

Comment (topperc) [reply llvm/lib/Transforms/Scalar/Reassociate.cpp:2258]: I looked at this some more. When we reassociate a set of Ors we lose the flag, but then we revisit them to try more reassociation that may want to treat them as Adds. But we've lost the flag. So I think we need to call haveNoCommonBitsSet again somewhere to recover it.

Comment (nikic) [reply llvm/lib/Transforms/Scalar/Reassociate.cpp:2258]: Okay, in that case keeping the haveNoCommonBitsSet calls makes sense.

**Review by preames — APPROVED** [review]
LGTM

**Pull request merged**